packages/relay/src/relay.ts

```diff
@@ -48,648 +33,129 @@ export interface RelayOptions {
-          },
-        );
-      })
-      .on(Announce, async (ctx, announceActivity) => {
-        const sender = await announceActivity.getActor(ctx);
-        const excludeBaseUris = sender?.id ? [new URL(sender.id)] : [];
-
-        const announce = new Announce({
-          id: new URL(`/announce#${crypto.randomUUID()}`, ctx.origin),
-          actor: ctx.getActorUri(RELAY_SERVER_ACTOR),
-          object: announceActivity.objectId,
-          to: PUBLIC_COLLECTION,
-        });
-
-        await ctx.sendActivity(
-          { identifier: RELAY_SERVER_ACTOR },
-          "followers",
-          announce,
-          {
-            excludeBaseUris,
-            preferSharedInbox: true,
-          },
-        );
-      });
-  }
-
-  get domain(): string {
-    return this.#options.domain || "localhost";
-  }
-
-  fetch(request: Request): Promise<Response> {
-    return this.#federation.fetch(request, { contextData: undefined });
-  }
 
-  setSubscriptionHandler(handler: SubscriptionRequestHandler): this {
-    this.#subscriptionHandler = handler;
-    return this;
+        return [rsaPair, ed25519Pair];
+      }
+
+      const rsaPair: CryptoKeyPair = {
+        privateKey: await importJwk(rsaPairJson.privateKey, "private"),
+        publicKey: await importJwk(rsaPairJson.publicKey, "public"),
+      };
+      const ed25519Pair: CryptoKeyPair = {
+        privateKey: await importJwk(ed25519PairJson.privateKey, "private"),
+        publicKey: await importJwk(ed25519PairJson.publicKey, "public"),
+      };
+      return [rsaPair, ed25519Pair];
+    },
+  );
+
+relayBuilder.setFollowersDispatcher(
+  "/users/{identifier}/followers",
+  async (ctx, identifier) => {
+    if (identifier !== RELAY_SERVER_ACTOR) return null;
+
+    const followers = await ctx.data.kv.get<string[]>(["followers"]) ??
+      [];
+
+    const actors: Actor[] = [];
+    for (const followerId of followers) {
+      const follower = await ctx.data.kv.get<RelayFollower>([
+        "follower",
+        followerId,
+      ]);
+      if (!follower) continue;
+      const actor = await Object.fromJsonLd(follower.actor);
+      if (!isActor(actor)) continue;
+      actors.push(actor);
+    }
+    return { items: actors };
+  },
+);
+
+relayBuilder.setFollowingDispatcher(
+  "/users/{identifier}/following",
+  async (ctx, identifier) => {
+    if (identifier !== RELAY_SERVER_ACTOR) return null;
+
+    const followers = await ctx.data.kv.get<string[]>(["followers"]) ??
+      [];
+
+    const actors: Actor[] = [];
+    for (const followerId of followers) {
+      const follower = await ctx.data.kv.get<RelayFollower>([
+        "follower",
+        followerId,
+      ]);
+      if (!follower) continue;
+      const actor = await Object.fromJsonLd(follower.actor);
+      if (!isActor(actor)) continue;
+      actors.push(actor);
+    }
+    return { items: actors };
+  },
+);
+
+export function createRelay(
+  type: string,
+  options: RelayOptions,
+): MastodonRelay | LitePubRelay {
```

Comment:
Rather than making its return type a union of two concrete types, it's much better to have a common interface for both concrete types, e.g., `Relay`, and making it return an abstract `Relay`.  Actually, it's why we have factory functions: to hide concrete types from outside and abstract away them.